### ISSUE TITLE
ORA-64451: Conversion of special character to escaped character failed.

### ISSUE THREAD

**Issue by Spamme1** [schemaspy/schemaspy#740]
I used SchemaSpy with an Oracle database with German names  (tables, columns, views, stored procedures, etc.)
During the execution SchemaSpy shown a warning:

> WARN  - Failed to retrieve stored procedure/function details using sql 'SELECT s.owner || '.' || s.name AS routine_name, s.TYPE AS routine_type, a.DATA_TYPE AS dtd_identifier, 'PL/SQL' AS routine_body, dbms_xmlgen.convert(xmlagg(xmlelement(e,s.text,'').extract('//text()') order by s.line).GetClobVal(), 1) AS routine_definition, null as is_deterministic, null AS sql_data_access, null AS security_type, null AS sql_mode, null AS routine_comment  FROM all_source s LEFT OUTER JOIN all_arguments a ON a.OWNER = s.owner AND a.OBJECT_NAME = s.NAME AND a.DATA_LEVEL = 0 AND a.argument_name IS null where s.owner = :schema GROUP BY s.owner, s.name, s.TYPE, a.data_type'
> ORA-64451: Conversion of special character to escaped character failed.

## Expected Behavior
SchemaSpy finished the generation of the HTML pages but I don't know if something is missing, It seems from the error that the stored procedures/functions couldn't have been read from the database because the error.

## Current Behavior
SchemaSpy finished the generation of HTML but I don't know if everything was generated correctly.

## Possible Solution
No idea, try to escape/remove special chars?
The German language can have this special chars: ü, ö, ä

## Steps to Reproduce (for bugs)
I can't attach the database because it contains customer data.

## Context
* Database type (e.g. MySQL): Oracle (v. 12)
* Database driver version (is you use a special one): latest version
* Your DDL (sample how to reproduce the issue): 
* Full command including arguments how you run SchemaSpy: java -jar schemaspy-6.1.0.jar -t orathin -dp ojdbc8-full -host oravm -port 1521 -db xe

## Your Environment
* Version used: 6.1.0
* Environment name and version (e.g. Chrome 39): 
* Operating System and version (e.g. Windows 10): Windows 10 Pro
* Java version: 8 update 261

**Comment by npetzall** [comment]
Since it's details, you'll probably end up with routine that goes to a empty och not existing page.

**Comment by npetzall** [comment]
https://stackoverflow.com/questions/41749765/ora-64451-conversion-of-special-character-to-escaped-character-failed

Found that which isn't of much help. Since batch get all details it's har do say which one has some Un-convertable character.

**Comment by eric-gagnon** [comment]
Hi! Stumbled on the exact same problem today.

I'm evaluating how I could use Schema Spy to understand / document / validate implementation of a complex functionnality in our application.

I discovered that there was some sort of support of trigger export but there was nothing visible in my own export (empty routines.html file).

I discovered the query config and tried to run it in SQLDeveloper to understand what it was doing. I'm not a Oracle specialist but understand after reading a bit that it's a common way to do it (instead of using listaggr).
Found something about xmlserialise but was not able to make it work.

Not sure what caracters are problematic but I tried different things and one thing that seem to work (but could require appropriate grant) is using DBMS_METADATA.

`DBMS_METADATA.get_ddl (s.TYPE,  s.name) AS routine`

In my case, I also had to filter out things in the bin folder (I had no idea what those script where comming from, just found an article explaining it (https://randallsims.wordpress.com/2015/06/29/oracle-recycle-bin-bin-and-rb-files/).

So added a AND s.name not like 'BIN% to filter those out (otherwise it generate an error of get_dll that cannot found those).

So the query that I run looks like this (in SQLDeveloper, I willl try to update the configuration and give it a try).

`SELECT s.owner || '.' || s.name AS routine_name, s.TYPE AS routine_type, a.DATA_TYPE AS dtd_identifier, 'PL/SQL' AS routine_body, DBMS_METADATA.get_ddl (s.TYPE,  s.name) AS routine_definition, null as is_deterministic, null AS sql_data_access, null AS security_type, null AS sql_mode, null AS routine_comment  FROM all_source s LEFT OUTER JOIN all_arguments a ON a.OWNER = s.owner AND a.OBJECT_NAME = s.NAME AND a.DATA_LEVEL = 0 AND a.argument_name IS null where s.owner = 'MY_SCHEMA'  AND s.name not like 'BIN%' GROUP BY s.owner, s.name, s.TYPE, a.data_type`

This solution will be likely good enough for me but it very likely not the best for the default configs.

ps: I discovered ShemaSpy last week by accident, try to understand a a schema by reading list of table name is a pain and SQLDevelopper generation is what it is. SchemaSPY a real GEM (it was a real pleasure to discover an xml output and also the intermediary .dot files) and I applaud the decision to keep it alive!

**Comment by npetzall** [comment]
@eric-gagnon

To have schemaspy use your query.

Create a file in the same folder as the jar named `myora.properties`

Should have content like
```
extends=orathin
# if orathin is your -t arg
selectRoutinesSql=SELECT s.owner || '.' || s.name AS routine_name, s.TYPE AS routine_type, a.DATA_TYPE AS dtd_identifier, 'PL/SQL' AS routine_body, DBMS_METADATA.get_ddl (s.TYPE, s.name) AS routine_definition, null as is_deterministic, null AS sql_data_access, null AS security_type, null AS sql_mode, null AS routine_comment FROM all_source s LEFT OUTER JOIN all_arguments a ON a.OWNER = s.owner AND a.OBJECT_NAME = s.NAME AND a.DATA_LEVEL = 0 AND a.argument_name IS null where s.owner = :schema AND s.name not like 'BIN%' GROUP BY s.owner, s.name, s.TYPE, a.data_type
```

No just switch `-t` arg to myora i.e `-t myora`

**Comment by npetzall** [comment]
https://docs.oracle.com/database/121/ARPLS/d_metada.htm#ARPLS026

**Comment by eric-gagnon** [comment]
Sorry, took a bit of time to get back to this.

Thank you very much for the exemple of override, just had to change the extends to orathin-service as this is what I'm using.

Changed back "MY_SCHEMA" reference of the query sample I posted to the binding variable :shema and everything went well  and I can see the triggers in routines tab.

**Comment by eric-gagnon** [comment]
It seems like the query is not fully right, I double checked there are a few triggers missing and checking again I do see an error in sqldeveloper when I scroll in the results (that was likely not present in the first 50 rows of my data).

`ORA-31600: valeur d'entrée PACKAGE BODY non valide pour le paramètre OBJECT_TYPE dans la fonction GET_DDL
ORA-06512: à "SYS.DBMS_METADATA", ligne 6478
ORA-06512: à "SYS.DBMS_METADATA", ligne 6264
ORA-06512: à "SYS.DBMS_SYS_ERROR", ligne 105
ORA-06512: à "SYS.DBMS_METADATA_INT", ligne 4152
ORA-06512: à "SYS.DBMS_METADATA_INT", ligne 4629
ORA-06512: à "SYS.DBMS_METADATA_INT", ligne 9617
ORA-06512: à "SYS.DBMS_METADATA", ligne 6207
ORA-06512: à "SYS.DBMS_METADATA", ligne 6320
ORA-06512: à "SYS.DBMS_METADATA", ligne 9202
ORA-06512: à ligne 1
31600. 00000 -  "invalid input value %s for parameter %s in function %s"
*Cause:    A NULL or invalid value was supplied for the parameter.
*Action:   Correct the input value and try the call again.`

I'll repost / share if I find a better version

**Comment by eric-gagnon** [comment]
Just a quick fix following hint of error but I have not taken the time to validate completely :
SELECT s.owner || '.' || s.name AS routine_name, s.TYPE AS routine_type, a.DATA_TYPE AS dtd_identifier, 'PL/SQL' AS routine_body, DBMS_METADATA.get_ddl (s.TYPE, s.name) AS routine_definition, null as is_deterministic, null AS sql_data_access, null AS security_type, null AS sql_mode, null AS routine_comment FROM all_source s LEFT OUTER JOIN all_arguments a ON a.OWNER = s.owner AND a.OBJECT_NAME = s.NAME AND a.DATA_LEVEL = 0 AND a.argument_name IS null where s.owner = 'GALA' AND s.name not like 'BIN%' **and s.type not like 'PACKAGE BODY' GROUP** BY s.owner, s.name, s.TYPE, a.data_type

**Comment by zpmartins** [comment]
In what concerns the previous error.... (ORA-31600)...

You just need to add to the WHERE clause... " AND s.type != 'PACKAGE BODY' " after the "... not like 'BIN%'"

The get_ddl procedure does not support that object type.

KR,

**Comment by eric-gagnon** [comment]
Yes, this is what I have discovered, I forgot to post an update sorry.
